src/terrain/terrain.js

```diff
@@ -294,13 +302,23 @@ export class Terrain extends Elevation {
             this._findCoveringTileCache[this.proxySourceCache.id] = {};
             this.proxySourceCache.update(transform);
 
-            this._depthDone = false;
             this._emptyDEMTextureDirty = true;
         } else {
             this._disable();
         }
     }
 
+    _checkRenderCacheEfficiency() {
+        const renderCacheInfo = this.renderCacheEfficiency(this._style);
+        if (this._style.map._optimizeForTerrain) {
+            assert(renderCacheInfo.efficiency === 100);
+        } else if (renderCacheInfo.efficiency !== 100) {
+            warnOnce(`Terrain render cache efficiency is not optimal (${renderCacheInfo.efficiency}%) and
+                performance may be affected negatively, consider adding draped layers before layer with id
```

Comment:
`draped layers` may not be clear for user, how about `consider placing all background, fill and line layers before..`

Reply:
Updated in https://github.com/mapbox/mapbox-gl-js/pull/10258/commits/b732630d4465ce232a58c71932843833dc2885e9.